Objects/typeobject.c

```diff
@@ -8248,14 +8248,17 @@ _Py_slot_tp_getattr_hook(PyObject *self, PyObject *name)
         (Py_IS_TYPE(getattribute, &PyWrapperDescr_Type) &&
          ((PyWrapperDescrObject *)getattribute)->d_wrapped ==
          (void *)PyObject_GenericGetAttr))
-        res = PyObject_GenericGetAttr(self, name);
+        /* finding nothing is reasonable when __getattr__ is defined */
+        res = _PyObject_GenericTryGetAttr(self, name);
     else {
         Py_INCREF(getattribute);
         res = call_attribute(self, getattribute, name);
         Py_DECREF(getattribute);
     }
-    if (res == NULL && PyErr_ExceptionMatches(PyExc_AttributeError)) {
-        PyErr_Clear();
+    if (res == NULL) {
+        if (PyErr_ExceptionMatches(PyExc_AttributeError)) {
+            PyErr_Clear();
+        }
```

Comment:
This causes `res` to fall through when NULL even when there's no `PyExc_AttributeError`. Is the behavior change intended? If so could you please explain it as I'm not quite understanding it?

Reply:
This is my understanding. The old code `res == NULL && PyErr_ExceptionMatches(PyExc_AttributeError)` has two behaviors:
1. Check res is not found when NULL is returned as a result.
2. Exceptions are cleared because both `PyObject_GenericGetAttr` and custom defined `__getattribute__` may throw exceptions.

`_PyObject_GenericTryGetAttr` may not raise exceptions any more, but `__getattribute__` may still has this behavior. So, if we delete
```
   if (PyErr_ExceptionMatches(PyExc_AttributeError)) {
      PyErr_Clear();
   }
```
we may end up with exceptions being raised:
```
class AttrObj(object):
    def __init__(self) -> None:
        super(AttrObj, self).__init__()
        self.pps = 2
        self.ppd = None

    def __getattr__(self, name):
        return 4

    def __getattribute__(self, name):
        if name == 'ppa':
            return 3

        raise AttributeError('no attribute %s'%name)

>>> from test_getattr import AttrObj as ao
>>> a = ao()
>>> a.ppa
3
>>> a.ppb
Traceback (most recent call last):
  File "/home/wangxiang/test_codes/test_getattr.py", line 19, in __getattribute__
    raise AttributeError('no attribute %s'%name)
AttributeError: no attribute ppb

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
SystemError: <function AttrObj.__getattr__ at 0x7fe459bef4c0> returned a result with an exception set
```

But strictly speaking, the completely equivalent expression might be:
```
    if (res == NULL) {
        if (getattribute != NULL && PyErr_ExceptionMatches(PyExc_AttributeError)) {
            PyErr_Clear();
        }
        res = call_attribute(self, getattr, name);
    }
```

I simplified it. It looks concise, probably no risk, but indeed result in an extra call of `PyErr_Clear` when `__getattribute__` is not defined.
Thank you.

Reply:
Nice work! Should your AttrObj be added as a test case?

Reply:
> Nice work! Should your AttrObj be added as a test case?

Thanks!
It's easy, I can organize and submit this test cast if it is helpful.

Reply:
> `_PyObject_GenericTryGetAttr` may not raise exceptions any more

This is not true. This misunderstanding caused trouble when accessing a property that raises exceptions. 
See #103272 #103329

Reply:
I think you are right. `_PyObject_GenericTryGetAttr` doesn't raise `PyExc_AttributeError` but other exceptions may alse be raised, that will cause `getattr` be called.
Thanks for finding this error. Later modifications you committed is better.